Demangle function names

data-pup opened the pull request:
Fixes #174. (Finally! :smile_cat:)

In #270, we [proposed](https://github.com/rustwasm/twiggy/pull/270#pullrequestreview-216947451) [merging](https://github.com/rustwasm/twiggy/pull/270#issuecomment-475041468) the IR items for code/function section entries.

>  I wonder if it makes sense to collapse func[123] and code[123] into the same IR item, since no body is used by more than one func and vice versa. They are essentially the same, even if they are split at the physical binary level (to allow eager type checking and validation).

> I think I agree [..] that the best way to handle this on WebAssembly is to merge the func/code nodes. When looking at the size of a function we may as well account for the bytes that the function type pointer takes up (although it's like 1-3 at most realistically). It'll definitely make reading the output a bit easier as I know in the past I've had to follow chains a few levels deep to get to the real part of a function!

This combines the func/code nodes in the IR generated when parsing a wasm module :tada: 

To keep this somewhat more reviewable, I've divided the PR into 3 commits.
*  9e82d6c - No changes to behavior, this simply refactors the parsing code towards a structure where merging the func/code nodes can be accomplished in as small a change as possible. This commit will still pass the original test snapshots.
*  474b193 - This is where the actual IR node merge occurs, including changes to how we draw edges between nodes.
*  e1fcfba - This commit is purely updating the test snapshots. This should probably be reviewed _first_, so that we can decide whether this is an acceptable change regarding ergonomics/correctness :slightly_smiling_face:

data-pup commented on twiggy/tests/all/expectations/dominators_wee_alloc_with_depth_and_row:
```diff
@@ -1,6 +1,6 @@
-            387 ┊     13.74% ┊ func[2]
-            378 ┊     13.42% ┊   ⤷ wee_alloc::alloc_with_refill::hb32c1bbce9ebda8e
+            387 ┊     13.74% ┊ wee_alloc::alloc_with_refill::hb32c1bbce9ebda8e
+            226 ┊      8.02% ┊   ⤷ wee_alloc::alloc_first_fit::h9a72de3af77ef93f
```
Here's the test case that I think demonstrates this change most succinctly. Now, the topmost node contains the name of the function/code. The immediate child is another function that this dominates, as opposed to the code node that was displayed before.

data-pup commented on twiggy/tests/all/expectations/garbage:
```diff
@@ -1,12 +1,9 @@
-    11 ┊  5.58% ┊ unusedAddThreeNumbers
-     8 ┊  4.06% ┊ unusedAddOne
+    12 ┊  6.09% ┊ unusedAddThreeNumbers
+     9 ┊  4.57% ┊ unusedAddOne
```
In general, most of the test changes look like this: A single byte being added to the node that now represents both the code _and_ function entries. Other changes, such as the rest of this file, are because there are fewer items listed in the 'total rows' message.

data-pup commented on twiggy/tests/all/expectations/paths_test_called_twice:
```diff
@@ -1,13 +1,9 @@
+               ┊           ┊       ⬑ awoo
+               ┊           ┊           ⬑ export "awoo"
+               ┊           ┊   ⬑ woof
+               ┊           ┊       ⬑ export "woof"
```
Paths outputs also look much nicer now! Fewer intermediate func[i] nodes.

data-pup commented on parser/wasm_parse/mod.rs:
```diff
@@ -73,81 +92,77 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
-                        .get_function_section_reader()?
-                        .parse_items(items, idx)?;
-                    "function section headers".to_string()
+                    panic!("unexpected function section");
```
tangentially, not _entirely_ sure what to do right here. We shouldn't ever hit this branch.

fitzgen replied:
If this branch is actually unreachable, as opposed to something that can happen but is an error, then we should use the `unreachable!()` macro. I believe that is the case here, right?

data-pup commented on parser/wasm_parse/mod.rs:
```diff
@@ -292,6 +325,176 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
+    }
+}
+
+impl<'a> Parse<'a> for (IndexedSection<'a>, IndexedSection<'a>) {
```
Function/Code sections are merged by parsing them together. This is where most of the important changes to this file are happening.

data-pup commented:
At the moment, there's still some light cleanup that I'd want to do before this gets merged. However, I wanted to open this up for review to check that the alterations to the output, and general approach to implementing this change, look reasonable :slightly_smiling_face:

data-pup commented on twiggy/tests/all/expectations/dominators_regex_any_func:
```diff
@@ -1,14 +1,2 @@
-              5 ┊      3.47% ┊   ⤷ bark
-              6 ┊      4.17% ┊ func[4]
-              5 ┊      3.47% ┊   ⤷ awoo
+────────────────┼────────────┼───────────────
```
this test would need to be updated to use a different regex, since `func[i]` isn't actually a name that shows up anymore!

fitzgen commented on parser/wasm_parse/mod.rs:
```diff
@@ -73,81 +92,77 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
+        match (function_section, code_section) {
+            (Some(function_section), Some(code_section)) => (function_section, code_section)
+                .parse_items(items, (imported_functions, &names, &sizes))?,
+            _ => panic!("function or code section is missing"),
```
This should be a `failure::bail!("...")` instead of a panic

fitzgen commented on parser/wasm_parse/mod.rs:
```diff
@@ -73,81 +92,77 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
-                        .get_code_section_reader()?
-                        .parse_items(items, (idx, imported_functions, &names))?;
-                    "code section headers".to_string()
+                    panic!("unexpected code section");
```
Ditto.

fitzgen commented on parser/wasm_parse/mod.rs:
```diff
@@ -274,10 +312,7 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
-                    section
-                        .get_code_section_reader()?
-                        .parse_edges(items, (&indices, idx))?;
+                    panic!("unexpected code section found");
```
These should be `unreachable!()` too, I think

fitzgen commented on parser/wasm_parse/mod.rs:
```diff
@@ -73,7 +73,9 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
-            _ => panic!("function or code section is missing"),
+            _ => Err(traits::Error::with_msg(
+                "function or code section is missing",
+            ))?,
```
Ah ok perfect -- I think `failure::bail!` wouldn't work for us anyways since I forgot that we don't use `failure::Error`

fitzgen commented on parser/wasm_parse/mod.rs:
```diff
@@ -122,15 +121,18 @@ impl<'a> Parse<'a> for wasmparser::ModuleReader<'a> {
+                    Err(traits::Error::with_msg("unexpected code section"))?
+                }
+                wasmparser::SectionCode::Function => {
+                    Err(traits::Error::with_msg("unexpected function section"))?
```
These should still be `unreachable!()` however.

data-pup commented:
Thanks for the review! I resolved your notes, adding `unreachable!` macros to branches that are in fact unreachable, addressed warnings from `clippy`, and rebased this off of current `master`. Once CI finishes up, this should be good to merge :cowboy_hat_face:

fitzgen commented:
Woo! Thanks @data-pup! I think we should get this in folks' hands -- is there anything else you want to fix before we do the next release?

data-pup commented:
No problem @fitzgen! Was fun to figure this one out. There isn't anything pressing enough atm to warrant holding off on a release, I think this provides enough improvement to the output that it's worth doing a release sooner than later.

fitzgen commented:
Published!

* https://users.rust-lang.org/t/announcing-twiggy-version-0-6-0/29708
* https://www.reddit.com/r/rust/comments/c6apb2/announcing_twiggy_version_060/
* https://twitter.com/fitzgen/status/1144344943554519040
